### ISSUE TITLE
Magento\Eav\Model\ResourceModel\ReadHandler wrong sort of entityData

### ISSUE THREAD

**Issue by schnappenberger** [magento/magento2#6958]
In the execute function of Magento\Eav\Model\ResourceModel\ReadHandler the attribute values were sort by store_id desc and then merged by a union query. In the following loop the store view values can be overwritten by default store value.
### Preconditions
- PHP 7
- Apache 2
- Mysql 5
- Magento 2.1.1
### Steps to reproduce
1. Create an attribute value_id for a store_view lower than value_id of store_view 0 for attribute "name" of product (Happened in a live environment)
2. Edit Product in backend and switch store_view from default to specific store_view
### Expected result
1. Name should contain correct translated attribute value
### Actual result
1. Name contains attribute value of default store_view

I think the order of the selects is not relevant and the whole union query should be sorted by store_id asc. In the following loop the values of store_view 0 are placed first and would then be overwritten by store specific attribute values.

**Comment by dima-welance** [comment]
I faced the same issue.
And tried to change the sorting , it looks like when using union in the way the query is built the sorting plays no role.

**Comment by dima-welance** [comment]
The issue also exists in Magento\Catalog\Model\ResourceModel\AbstractResource _getLoadAttributesSelect

**Comment by dimonixx** [comment]
In Magento\Catalog\Model\ResourceModel\AbstractResource the issue could be solved by adding an additional select and return it. 

`$orderedSelect = $this->getConnection()->select()->from(['attr_values_sorted' => $select])->columns('*');`

The method would look like this:

```
protected function _getLoadAttributesSelect($object, $table)
    {
        if ($this->storeManager->hasSingleStore()) {
            $storeId = (int)$this->storeManager->getStore(true)->getId();
        } else {
            $storeId = (int)$object->getStoreId();
        }
        $setId = $object->getAttributeSetId();
        $storeIds = [$this->getDefaultStoreId()];
        if ($storeId != $this->getDefaultStoreId()) {
            $storeIds[] = $storeId;
        }

        $select = $this->getConnection()
            ->select()
            ->from(['attr_table' => $table], [])
            ->columns('*')
            ->where("attr_table.{$this->getEntityIdField()} = ?", $object->getId())
            ->where('attr_table.store_id IN (?)', $storeIds);

        if ($setId) {
            $select->join(
                ['set_table' => $this->getTable('eav_entity_attribute')],
                $this->getConnection()->quoteInto(
                    'attr_table.attribute_id = set_table.attribute_id' . ' AND set_table.attribute_set_id = ?',
                    $setId
                ),
                []
            );
        }
        $select->order('attr_table.store_id ASC');
        $orderedSelect = $this->getConnection()->select()->from(['attr_values_sorted' => $select])->columns('*');
        return $orderedSelect;
    }
```
Performance impacts were not tested.

**Comment by magento-engcom-team** [comment]
@schnappenberger, thank you for your report.
We've created internal ticket(s) MAGETWO-82301 to track progress on the issue.

**Comment by marcinfr** [comment]
Seems that SQL UNION operator corrupts sort order of selects result. 
I have found solution for this issue, See referenced PR #13752 .

**Comment by kanduvisla** [comment]
What's the priority on this issue? Because since Magento 2.2.x the entity manager has gotten a much bigger role when it comes to handling products.

**Comment by taimaiduc** [comment]
I have the same issue on #14221

**Comment by domeglic** [comment]
Looks like this has been fixed in [2.3](https://github.com/magento/magento2/commits/2.3-develop/app/code/Magento/Eav/Model/ResourceModel/ReadHandler.php). Commit message mentions sorting. (MAGETWO-95827)
